broker/src/main/java/org/apache/rocketmq/broker/pop/PopConsumerService.java

```diff
@@ -0,0 +1,699 @@
+                // should not add accumulation(max offset - consumer offset) here
+                return CompletableFuture.completedFuture(result);
+            }
+
+            int remain = batchSize - result.getMessageCount();
+            if (remain <= 0) {
+                result.addRestCount(this.getPendingFilterCount(groupId, topicId, queueId));
+                return CompletableFuture.completedFuture(result);
+            } else {
+                long consumeOffset = brokerController.getConsumerOffsetManager().queryPullOffset(groupId, topicId, queueId);
+                return getMessageAsync(clientHost, groupId, topicId, queueId, consumeOffset, remain, filter)
+                    .thenApply(getMessageResult -> addGetMessageResult(
+                        result, getMessageResult, topicId, queueId, retryType, consumeOffset));
+            }
+        });
+    }
+
+    public CompletableFuture<PopConsumerContext> popAsync(String clientHost, long popTime, long invisibleTime,
+        String groupId, String topicId, int queueId, int batchSize, boolean fifo, String attemptId,
+        MessageFilter filter) {
+
+        PopConsumerContext popConsumerContext =
+            new PopConsumerContext(clientHost, popTime, invisibleTime, groupId, fifo, attemptId);
+
+        TopicConfig topicConfig = brokerController.getTopicConfigManager().selectTopicConfig(topicId);
+        if (topicConfig == null || !consumerLockService.tryLock(groupId, topicId)) {
+            return CompletableFuture.completedFuture(popConsumerContext);
+        }
+
+        log.debug("PopConsumerService popAsync, groupId={}, topicId={}, queueId={}, " +
+                "batchSize={}, invisibleTime={}, fifo={}, attemptId={}, filter={}",
+            groupId, topicId, queueId, batchSize, invisibleTime, fifo, attemptId, filter);
+
+        String requestKey = groupId + "@" + topicId;
+        String retryTopicV1 = KeyBuilder.buildPopRetryTopicV1(topicId, groupId);
+        String retryTopicV2 = KeyBuilder.buildPopRetryTopicV2(topicId, groupId);
+        long requestCount = Objects.requireNonNull(ConcurrentHashMapUtils.computeIfAbsent(
+            requestCountTable, requestKey, k -> new AtomicLong(0L))).getAndIncrement();
+        boolean preferRetry = requestCount % 5L == 0L;
+
+        CompletableFuture<PopConsumerContext> getMessageFuture =
+            CompletableFuture.completedFuture(popConsumerContext);
+
+        try {
+            GetMessageResult finalGetMessageResult = new GetMessageResult();
+            finalGetMessageResult.setStatus(GetMessageStatus.NO_MESSAGE_IN_QUEUE);
+
+            if (!fifo && preferRetry) {
+                if (brokerConfig.isRetrieveMessageFromPopRetryTopicV1()) {
+                    getMessageFuture = this.getMessageAsync(getMessageFuture, clientHost, groupId,
+                        retryTopicV1, 0, batchSize, filter, PopConsumerRecord.RetryType.RETRY_TOPIC_V1);
```

Comment:
Here isRetrieveMessageFromPopRetryTopicV1 and isEnableRetryTopicV2 may be established at the same time in compatible scenarios. At this time, when processing getMessageFuture here, will it always use isEnableRetryTopicV2?

Reply:
In the final version, the v2 format is used only. During the retry topic format transfer phase, there may be unconsumed retry messages in the v1 topic.